tests/test_experimental.py

```diff
@@ -630,6 +632,330 @@ def multiply(self, X, K):
+    X_prior: npt.NDArray[np.float64],
+    obs_index_flatten: int,
+    rho: npt.NDArray[np.float64],
+    true_parameters: npt.NDArray[np.float64],
+    X_posterior: npt.NDArray[np.float64],
+    X_posterior_global: npt.NDArray[np.float64],
+    rho_min: float = 1e-6,
+    atol: float = 1e-5,
+):
+    # Find parameters far from the observation where localization weight is near zero
+    zero_weight_indices = np.where(rho < rho_min)[0]
+    assert len(zero_weight_indices) > 0, (
+        "No distant points found for testing localization."
+    )
+    # --- Assertions for Localized Smoother ---
+    # Assert distant parameters are not updated
+    assert np.allclose(
+        X_posterior[zero_weight_indices, :], X_prior[zero_weight_indices, :], atol=atol
+    )
+    # Assert parameter at observation IS updated
+    assert not np.allclose(
+        X_posterior[obs_index_flatten, :], X_prior[obs_index_flatten, :]
+    )
+
+    prior_mean = np.mean(X_prior, axis=1)
+    posterior_mean = np.mean(X_posterior, axis=1)
+    prior_variance = np.var(X_prior, axis=1)
+    posterior_variance = np.var(X_posterior, axis=1)
+
+    # Assert update is maximal at the observation point
+    update_magnitude = np.abs(posterior_mean - prior_mean)
+    assert np.argmax(update_magnitude) == obs_index_flatten
+
+    # Assert variance is reduced at observation, but preserved far away
+    assert posterior_variance[obs_index_flatten] < prior_variance[obs_index_flatten]
+    # The variance of the ensemble should NOT change where there was no update
+    assert np.allclose(
+        posterior_variance[zero_weight_indices], prior_variance[zero_weight_indices]
+    )
+
+    # The global update MUST change distant parameters due to spurious correlations.
+    assert not np.allclose(
+        X_posterior_global[zero_weight_indices, :],
+        X_prior[zero_weight_indices, :],
+    )
+
+    # --- Comparison Assertions for Global Smoother ---
+    posterior_mean_global = np.mean(X_posterior_global, axis=1)
+    posterior_variance_global = np.var(X_posterior_global, axis=1)
+
+    # Assert distant parameters ARE updated due to spurious correlations
+    assert not np.allclose(
+        X_posterior_global[zero_weight_indices, :],
+        X_prior[zero_weight_indices, :],
+    )
+
+    # Assert AVERAGE variance is reduced everywhere (the sign of ensemble collapse)
+    assert np.mean(posterior_variance_global[zero_weight_indices]) < np.mean(
+        prior_variance[zero_weight_indices]
+    )
+
+    # Calculate Mean Squared Error (MSE) against the true parameters
+    mse_localized = np.mean((posterior_mean - true_parameters) ** 2)
+    mse_non_localized = np.mean((posterior_mean_global - true_parameters) ** 2)
+
+    # --- Overall Quality Assertion (MSE) ---
+    # Assert that the localized result is a more plausible estimate
+    assert mse_localized < mse_non_localized
+
+
+def calculate_rho_1d(N_m: int, obs_index: int, localization_radius: float):
```

Comment:
nice! Since you already doing the type annotations, could you do the return type (for all rho calculations)?